File: src/main/java/xyz/mlhmz/savingscategorization/models/Pocket.java
package xyz.mlhmz.savingscategorization.models;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "pocket", indexes = @Index(columnList = "identifier"))
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Pocket {
    @Id
    @GeneratedValue(generator = "UuidGenerator2")
    private UUID uuid;
    private String identifier;
    private String name;
    private String description;
    private String iconName;
    @ElementCollection(targetClass = String.class, fetch = FetchType.EAGER)
    @CollectionTable(name = "pocket_keywords", joinColumns = @JoinColumn(name = "pocket_uuid"))
    @Column(name = "keyword", nullable = false)
    private List<String> keywords;
    private double transactionSum;
    @OneToMany(fetch = FetchType.LAZY)
    private List<Transaction> transactions;

    public Pocket(String name, String description, String iconName, List<String> keywords) {
        this.name = name;
        this.description = description;
        this.iconName = iconName;
        this.keywords = keywords;
    }
}

File: frontend/src/types/Pocket.ts
import { z } from "zod";
import { Transactions } from "./Transaction";

export interface Pocket {
  uuid?: string;
  name?: string;
  identifier?: string;
  description?: string;
  keywords?: string[];
  transactionSum?: number;
}

export const PocketMutation = z.object({
  name: z.string(),
  identifier: z.string(),
  description: z.string().optional(),
  keywords: z.string().array(),
});

export type PocketMutation = z.infer<typeof PocketMutation>;

export type PocketTransactions = {
  pocket?: Pocket;
  transactions?: Transactions;
};

export type Pockets = Array<Pocket>;
